quran/models.py

```diff
@@ -1,19 +1,88 @@
 '''Models describing abstractions used at Quran app'''
-from django.db import migrations, models
+from os import path
+from django.db import models
+from django.template.defaultfilters import slugify
 
 
 class Surah(models.Model):
-    '''General information about Surahs'''
+    '''Model representing a surah from the Qur’an.'''
     number = models.PositiveIntegerField(primary_key=True)
     name = models.CharField(max_length=20)
     total_ayahs = models.PositiveIntegerField()
 
+    def __str__(self):
+        return "%s:(%s)" % (self.name, self.number)
+
 
 class Ayah(models.Model):
-    '''General information about Ayah'''
+    '''Model representing a ayah from the Qur’an.'''
     number = models.PositiveIntegerField()
     text = models.TextField()
     surah = models.ForeignKey(
         Surah,
         on_delete=models.CASCADE,
         related_name='ayahs')
+
+    def __str__(self):
+        return "%s: (%s:%s)" % (
+            self.surah.name, self.surah.number, self.number)
+
+
+class Reciter(models.Model):
+    '''Model representing a reciter'''
+    name = models.CharField(max_length=100)
+    quality = models.CharField(max_length=20, blank=True)
+    style = models.CharField(max_length=100, blank=True)
+    slug = models.SlugField()
+
+    def save(self, *args, **kwargs):
+        self.quality = self.quality.replace('/', 'p').replace(' ', '')
+        self.slug = slugify(self.name)
+        super(Reciter, self).save(*args, **kwargs)
+
+    def __str__(self):
+        return self.name
+
+
+def audio_directory_path(instance, filename):
+    '''
+    the function generates path for
+    uploading audio file of ayahs to
+    MEDIA_ROOT/<reciter-slug>/<style>/<quality>/<surah-id>/<ayah-id>.mp3
+    '''
+    ayah = Ayah.objects.get(pk=instance.ayah_id)
+    surah_number = ayah.surah.number
+    file_extension = path.splitext(path.basename(filename))[1]
+    style_dir = path.join(
+        '/', instance.reciter.style
+    ) if instance.reciter.style else ''
+    quality_dir = path.join(
+        '/', instance.reciter.quality
+    ) if instance.reciter.quality else ''
+    return '{slug}{style}{quality}/{surah}/{ayah}.{extension}'.format(
+        slug=instance.reciter.slug,
+        style=style_dir,
+        quality=quality_dir,
+        surah=surah_number,
+        ayah=ayah.number,
+        extension=file_extension
+    )
+
+
+class Recitation(models.Model):
+    '''
+    This model represents a recitation of an ayah from the Qur'an.
+    Each recitation has an audio file and time segments which hold
+    an information about times when each word in this audio file
+    was pronounced.
+    '''
+    ayah_id = models.PositiveIntegerField()
+    segments = models.TextField()
```

Comment:
According to https://stackoverflow.com/questions/22340258/django-list-field-in-model you can redefine your own getter and setter for the field. I never tried it, but it looks very promising. Can you write your custom `getter` and `setter` for this field?
Test it and show that it works fine.

Reply:
Ok, but we should give to setter function list of segments. It will look like after dumping
'[[0,1,180,1230],[1,2,1240,2160],[2,3,2170,3040],[3,4,3050,4380],[4,5,4390,5410],[42,43,59530,62890]]'
And we should make something like [i[2:] for i in segments] before giving to setter function